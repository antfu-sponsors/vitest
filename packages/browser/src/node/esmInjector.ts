import MagicString from 'magic-string'
import { extract_names as extractNames } from 'periscopic'
import type { PluginContext } from 'rollup'
import { esmWalker } from '@vitest/utils/ast'
import type { Expression, ImportDeclaration, Node, Positioned } from '@vitest/utils/ast'

const viInjectedKey = '__vi_inject__'
// const viImportMetaKey = '__vi_import_meta__' // to allow overwrite
const viExportAllHelper = '__vitest_browser_runner__.exportAll'

const skipHijack = [
  '/@vite/client',
  '/@vite/env',
  /vite\/dist\/client/,
]

// this is basically copypaste from Vite SSR
// this method transforms all import and export statements into `__vi_injected__` variable
// to allow spying on them. this can be disabled by setting `slowHijackESM` to `false`
export function injectVitestModule(code: string, id: string, parse: PluginContext['parse']) {
  if (skipHijack.some(skip => id.match(skip)))
    return

  const s = new MagicString(code)

  let ast: any
  try {
    ast = parse(code)
  }
  catch (err) {
    console.error(`Cannot parse ${id}:\n${(err as any).message}`)
    return
  }

  let uid = 0
  const idToImportMap = new Map<string, string>()
  const declaredConst = new Set<string>()

  const hoistIndex = 0

  // this will transform import statements into dynamic ones, if there are imports
  // it will keep the import as is, if we don't need to mock anything
  // in browser environment it will wrap the module value with "vitest_wrap_module" function
  // that returns a proxy to the module so that named exports can be mocked
  const transformImportDeclaration = (node: ImportDeclaration) => {
    const source = node.source.value as string

    if (skipHijack.some(skip => source.match(skip)))
      return null

    const importId = `__vi_esm_${uid++}__`
    const hasSpecifiers = node.specifiers.length > 0
    const code = hasSpecifiers
      ? `import { ${viInjectedKey} as ${importId} } from '${source}'\n`
      : `import '${source}'\n`
    return {
      code,
      id: importId,
    }
  }

  function defineImport(node: ImportDeclaration) {
    const declaration = transformImportDeclaration(node)
    if (!declaration)
      return null
    s.appendLeft(hoistIndex, declaration.code)
    return declaration.id
  }

  function defineImportAll(source: string) {
    const importId = `__vi_esm_${uid++}__`
    s.appendLeft(hoistIndex, `const { ${viInjectedKey}: ${importId} } = await import(${JSON.stringify(source)});\n`)
    return importId
  }

  function defineExport(position: number, name: string, local = name) {
    s.appendLeft(
      position,
      `\nObject.defineProperty(${viInjectedKey}, "${name}", `
        + `{ enumerable: true, configurable: true, get(){ return ${local} }});`,
    )
  }

  // 1. check all import statements and record id -> importName map
  for (const node of ast.body as Node[]) {
    // import foo from 'foo' --> foo -> __import_foo__.default
    // import { baz } from 'foo' --> baz -> __import_foo__.baz
    // import * as ok from 'foo' --> ok -> __import_foo__
    if (node.type === 'ImportDeclaration') {
      const importId = defineImport(node)
      if (!importId)
        continue
      s.remove(node.start, node.end)
      for (const spec of node.specifiers) {
        if (spec.type === 'ImportSpecifier') {
          idToImportMap.set(
            spec.local.name,
            `${importId}.${spec.imported.name}`,
          )
        }
        else if (spec.type === 'ImportDefaultSpecifier') {
          idToImportMap.set(spec.local.name, `${importId}.default`)
        }
        else {
          // namespace specifier
          idToImportMap.set(spec.local.name, importId)
        }
      }
    }
  }

  // 2. check all export statements and define exports
  for (const node of ast.body as Node[]) {
    // named exports
    if (node.type === 'ExportNamedDeclaration') {
      if (node.declaration) {
        if (
          node.declaration.type === 'FunctionDeclaration'
          || node.declaration.type === 'ClassDeclaration'
        ) {
          // export function foo() {}
          defineExport(node.end, node.declaration.id!.name)
        }
        else {
          // export const foo = 1, bar = 2
          for (const declaration of node.declaration.declarations) {
            const names = extractNames(declaration.id as any)
            for (const name of names)
              defineExport(node.end, name)
          }
        }
        s.remove(node.start, (node.declaration as Node).start)
      }
      else {
        s.remove(node.start, node.end)
        if (node.source) {
          // export { foo, bar } from './foo'
          const importId = defineImportAll(node.source.value as string)
          // hoist re-exports near the defined import so they are immediately exported
          for (const spec of node.specifiers) {
            defineExport(
              hoistIndex,
              spec.exported.name,
              `${importId}.${spec.local.name}`,
            )
          }
        }
        else {
          // export { foo, bar }
          for (const spec of node.specifiers) {
            const local = spec.local.name
            const binding = idToImportMap.get(local)
            defineExport(node.end, spec.exported.name, binding || local)
          }
        }
      }
    }

    // default export
    if (node.type === 'ExportDefaultDeclaration') {
      const expressionTypes = ['FunctionExpression', 'ClassExpression']
      if (
        'id' in node.declaration
        && node.declaration.id
        && !expressionTypes.includes(node.declaration.type)
      ) {
        // named hoistable/class exports
        // export default function foo() {}
        // export default class A {}
        const { name } = node.declaration.id
        s.remove(node.start, node.start + 15 /* 'export default '.length */)
        s.append(
          `\nObject.defineProperty(${viInjectedKey}, "default", `
            + `{ enumerable: true, configurable: true, value: ${name} });`,
        )
      }
      else {
        // anonymous default exports
        s.update(
          node.start,
          node.start + 14 /* 'export default'.length */,
          `${viInjectedKey}.default =`,
        )
        // keep export default for optimized dependencies
        s.append(`\nexport default { ${viInjectedKey}: ${viInjectedKey}.default };\n`)
      }
    }

    // export * from './foo'
    if (node.type === 'ExportAllDeclaration') {
      s.remove(node.start, node.end)
      const importId = defineImportAll(node.source.value as string)
      // hoist re-exports near the defined import so they are immediately exported
      if (node.exported)
        defineExport(hoistIndex, node.exported.name, `${importId}`)
      else
        s.appendLeft(hoistIndex, `${viExportAllHelper}(${viInjectedKey}, ${importId});\n`)
    }
  }

  // 3. convert references to import bindings & import.meta references
  esmWalker(ast, {
    onIdentifier(id, info, parentStack) {
      const binding = idToImportMap.get(id.name)
      if (!binding)
        return

      if (info.hasBindingShortcut) {
        s.appendLeft(id.end, `: ${binding}`)
      }
      else if (
        info.classDeclaration
      ) {
        if (!declaredConst.has(id.name)) {
          declaredConst.add(id.name)
          // locate the top-most node containing the class declaration
          const topNode = parentStack[parentStack.length - 2]
          s.prependRight(topNode.start, `const ${id.name} = ${binding};\n`)
        }
      }
      else if (
        // don't transform class name identifier
        !info.classExpression
      ) {
        s.update(id.start, id.end, binding)
      }
    },
    // TODO: make env updatable
    onImportMeta() {
      // s.update(node.start, node.end, viImportMetaKey)
    },
    onDynamicImport(node) {
      const replace = '__vitest_browser_runner__.wrapModule(import('
      s.overwrite(node.start, (node.source as Positioned<Expression>).start, replace)
      s.overwrite(node.end - 1, node.end, '))')
    },
  })

  // make sure "__vi_injected__" is declared as soon as possible
  // prepend even if file doesn't export anything
  s.prepend(`const ${viInjectedKey} = { [Symbol.toStringTag]: "Module" };\n`)
  s.append(`\nexport { ${viInjectedKey} }`)

  return {
    ast,
    code: s.toString(),
    map: s.generateMap({ hires: 'boundary', source: id }),
  }
}
