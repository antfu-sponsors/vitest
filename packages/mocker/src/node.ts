export { automockModule, automockPlugin } from './plugin'
export { findMockRedirect } from './redirect'
