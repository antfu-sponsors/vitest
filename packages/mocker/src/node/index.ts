export { automockModule, automockPlugin } from './automockPlugin'
export { findMockRedirect } from './redirect'
export { hoistMocksPlugin, hoistMocks } from './hoistMocksPlugin'
export { ServerMockResolver } from './resolver'
export { dynamicImportPlugin } from './dynamicImportPlugin'
export { interceptorPlugin } from './interceptorPlugin'
export { mockerPlugin } from './mockerPlugin'

export type {
  ServerMockResolution,
  ServerIdResolution,
  ServerResolverOptions,
} from './resolver'
export type { AutomockPluginOptions } from './automockPlugin'
export type { HoistMocksPluginOptions, HoistMocksResult } from './hoistMocksPlugin'
export type { InterceptorPluginOptions } from './interceptorPlugin'
