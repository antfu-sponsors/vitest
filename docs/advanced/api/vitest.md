---
outline: deep
---

# Vitest

Vitest instance requires the current test mode. It can be either:

- `test` when running runtime tests
- `benchmark` when running benchmarks

## mode

### test

Test mode will only call functions inside `test` or `it`, and throws an error when `bench` is encountered. This mode uses `include` and `exclude` options in the config to find test files.

### benchmark

Benchmark mode calls `bench` functions and throws an error, when it encounters `test` or `it`. This mode uses `benchmark.include` and `benchmark.exclude` options in the config to find benchmark files.

## start

```ts
function start(filters: string[]): Promise<TestRunResult>
```

You can start running tests or benchmarks with `start` method. You can pass an array of strings to filter test files.

## config

The root (or global) config. If workspace feature is enabled, projects will reference this as `globalConfig`.

::: warning
This is Vitest config, it doesn't extend _Vite_ config. It only has resolved values from the `test` property.
:::

## vite

This is a global [`ViteDevServer`](https://vite.dev/guide/api-javascript#vitedevserver).

## state

::: warning
Public `state` is an experimental API. Breaking changes might not follow SemVer, please pin Vitest's version when using it.
:::

Global state stores information about the current tests. It uses the same API from `@vitest/runner` by default, but we recommend using the [Reported API](/advanced/reporters#reported-tasks) instead by calling `state.getReportedEntity()` on the `@vitest/runner` API:

```ts
const task = vitest.state.idMap.get(taskId) // old API
const testCase = vitest.state.getReportedEntity(task) // new API
```

In the future, the old API won't be exposed anymore.

## snapshot

The global snapshot manager. Vitest keeps track of all snapshots using the `snapshot.add` method.

You can get the latest summary of snapshots via the `vitest.snapshot.summay` property.

## cache

Cache manager that stores information about latest test results and test file stats. In Vitest itself this is only used by the default sequencer to sort tests.

## ready

Vitest needs to be resolved with the Vite server to be properly initialized. If the `Vitest` instance was created manually, you might need to check the `ready` status before accessing the `vite`, `state`, `cache`, `config`, and `snapshot` properties; otherwise, they will throw an error in the getter.

In normal circumstances, you would never call this method because `createVitest` and `startVitest` return already resolved Vitest instance.

## getRootTestProject

This returns the root test project. The root project generally doesn't run any tests and is not included in `vitest.projects` unless the user explicitly includes the root config in their workspace.

The primary goal of the root project is to setup the global config. In fact, `rootProject.config` references `rootProject.globalConfig` and `vitest.config` directly.

## provide

Vitest exposes `provide` method which is a shorthand for `vitest.getRootTestProject().provide`. With this method you can pass down values from the main thread to tests. All values are checked with `structuredClone` before they are stored, but the values themselves are not cloned.

To recieve the values in the test, you need to import `inject` method from `vitest` entrypont:

```ts
import { inject } from 'vitest'
const port = inject('wsPort') // 3000
```

For better type safety, we encourage you to augment the type of `ProvidedContext`:

```ts
import { createVitest } from 'vitest/node'

const vitest = await createVitest('test', {
  watch: false,
})
vitest.provide('wsPort', 3000)

declare module 'vitest' {
  export interface ProvidedContext {
    wsPort: number
  }
}
```

::: warning
Technically, `provide` is a method of [`TestProject`](#testproject), so it is limited to the specific project. However, all projects inherit the values from the core project which makes `vitest.provide` universal way of passing down values to tests.
:::

## getProvidedContext

This returns the root context object. This is a shorthand for `vitest.getRootTestProject().getProvidedContext`.

## getProjectByName

This method returns the project by its name. Simillar to calling `vitest.projects.find`.

::: warning
In case the project doesn't exist, this method will return the root project - make sure to check the names again if you need to make sure the project you are looking for is the one returned.
:::

## globTestSpecifications

This method constructs new [test specifications](#testspecification) by collecting every test in all projects with [`project.globTestFiles`](#globtestfiles). It accepts string filters to match the test files.

::: warning
As of Vitest 2.2.0, it's possible to have multiple test specifications with the same module ID (file path) if `poolMatchGlob` has several pools or if `typecheck` is enabled. This possibility will be removed in Vitest 3.
:::

```ts
const specifications = await vitest.globTestSpecifications(['my-filter'])
// [TestSpecification{ moduleId: '/tests/my-filter.test.ts' }]
console.log(specifications)
```

## mergeReports
## collect
## listFiles
## start
## init

## getModuleSpecifications

Returns a list of test specifications related to the module ID. The ID should already be resolved to an absolute file path. If ID doesn't match `include` or `includeSource` patterns, the returned array will be empty.

::: warning
As of Vitest 2.2.0, this method uses a cache to check if the file is a test. To make sure that the cache is not empty, call `globTestSpecifications` at least once.
:::

## runTestSpecifications
## rerunTestSpecifications
## collectTests
## cancelCurrentRun
## updateSnapshot
## invalidateFile
## close
## exit

## shouldKeepServer
## onServerRestart
## onCancel
## onClose
## onTestsRerun
## onFilterWatchedSpecification
