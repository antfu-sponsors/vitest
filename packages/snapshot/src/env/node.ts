import { existsSync, promises as fs } from 'node:fs'
import { basename, dirname, isAbsolute, join, resolve } from 'pathe'
import type { SnapshotEnvironment } from '../types'

export class NodeSnapshotEnvironment implements SnapshotEnvironment {
  getVersion(): string {
    return '1'
  }

  getHeader(): string {
    return `// Snapshot v${this.getVersion()}`
  }

  async resolveRawPath(testPath: string, rawPath: string) {
    return isAbsolute(rawPath)
      ? rawPath
      : resolve(dirname(testPath), rawPath)
  }

  async resolvePath(filepath: string): Promise<string> {
    return join(
      join(
        dirname(filepath), '__snapshots__'),
      `${basename(filepath)}.snap`,
    )
  }

  async prepareDirectory(filepath: string): Promise<void> {
    await fs.mkdir(filepath, { recursive: true })
  }

  async saveSnapshotFile(filepath: string, snapshot: string): Promise<void> {
    await fs.writeFile(filepath, snapshot, 'utf-8')
  }

  async readSnapshotFile(filepath: string): Promise<string | null> {
    if (!existsSync(filepath))
      return null
    return fs.readFile(filepath, 'utf-8')
  }

  async removeSnapshotFile(filepath: string): Promise<void> {
    if (existsSync(filepath))
      await fs.unlink(filepath)
  }
}
