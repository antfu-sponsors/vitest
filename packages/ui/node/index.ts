import { fileURLToPath } from 'node:url'
import { readFile } from 'node:fs/promises'
import { resolve } from 'pathe'
import sirv from 'sirv'
import type { Plugin } from 'vite'
import type { Vitest } from 'vitest'
import { resolveCoverageFolder, transformCoverageEntryPoint } from './utils'

export default (ctx: Vitest): Plugin => {
  return <Plugin>{
    name: 'vitest:ui',
    apply: 'serve',
    configureServer(server) {
      const uiOptions = ctx.config
      const base = uiOptions.uiBase
      const coverageFolder = resolveCoverageFolder(ctx)
      const coveragePath = coverageFolder ? coverageFolder[1] : undefined
      if (coveragePath && base === coveragePath) {
        throw new Error(
          `The ui base path and the coverage path cannot be the same: ${base}, change coverage.reportsDirectory`,
        )
      }

      if (coverageFolder) {
        const root = resolve(fileURLToPath(import.meta.url), '../')
        const base = readFile(resolve(root, '../istanbul-base.css'))
        const prettify = readFile(resolve(root, '../istanbul-prettify.css'))
        server.middlewares.use(coveragePath!, async (req, res, next) => {
          if (!req.url) {
            return next()
          }

          const pathname = new URL(req.url, 'http://localhost').pathname
          if (pathname !== '/' && pathname !== '/index.html' && pathname !== '/base.css' && pathname !== '/prettify.css' && !pathname.endsWith('.html')) {
            return next()
          }

          if (pathname === '/base.css') {
            res.setHeader('Content-Type', 'text/css')
            res.setHeader('Cache-Control', 'public,max-age=0,must-revalidate')
            res.end(await base)
            // for testing purposes: don't remove it
            // res.end(await readFile(resolve(fileURLToPath(import.meta.url), '../../istambul-base.css')))
            return
          }

          if (pathname === '/prettify.css') {
            res.setHeader('Content-Type', 'text/css')
            res.setHeader('Cache-Control', 'public,max-age=0,must-revalidate')
            res.end(await prettify)
            // for testing purposes: don't remove it
            // res.end(await readFile(resolve(fileURLToPath(import.meta.url), '../../istambul-prettify.css')))
            return
          }

          const file = pathname.endsWith('.html')
            ? pathname[0] === '/' ? pathname.slice(1) : pathname
            : 'index.html'

          const indexHtml = await readFile(resolve(coverageFolder[0], file), 'utf-8')
          res.setHeader('Content-Type', 'text/html')
          res.setHeader('Cache-Control', 'public,max-age=0,must-revalidate')
          res.end(transformCoverageEntryPoint(indexHtml))
        })
        server.middlewares.use(
          coveragePath!,
          sirv(coverageFolder[0], {
            single: true,
            dev: true,
            setHeaders: (res) => {
              res.setHeader(
                'Cache-Control',
                'public,max-age=0,must-revalidate',
              )
            },
          }),
        )
      }

      const clientDist = resolve(fileURLToPath(import.meta.url), '../client')
      server.middlewares.use(
        base,
        sirv(clientDist, {
          single: true,
          dev: true,
        }),
      )
    },
  }
}
