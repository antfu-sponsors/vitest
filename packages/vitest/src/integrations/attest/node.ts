import type { TestProject } from '../../node/project'
import { execFileSync } from 'node:child_process'
import { mkdirSync } from 'node:fs'

async function setup() {
  // TODO: programatic API
  mkdirSync('.attest/assertions', { recursive: true })
  execFileSync('attest', ['precache', '.attest/assertions/typescript.json'], {
    stdio: 'inherit',
    env: {
      ...process.env,
      ATTEST_attestAliases: JSON.stringify(['expect']),
    },
  })
}

export async function globalSetupAttest(project: TestProject) {
  // TODO: vitest level config
  // TODO: workspace support?
  if (!process.env.VITEST_ATTEST) {
    return
  }

  await setup()
  project.onTestsRerun(() => setup())
}
