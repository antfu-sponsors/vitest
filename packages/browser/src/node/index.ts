import type { WorkspaceProject } from 'vitest/node'
import type { Plugin } from 'vitest/config'
import { createViteLogger, createViteServer } from 'vitest/node'
import c from 'tinyrainbow'
import { version } from '../../package.json'
import { setupBrowserRpc } from './rpc'
import { BrowserServer } from './server'
import BrowserPlugin from './plugin'

export type { BrowserServer } from './server'
export { createBrowserPool } from './pool'

export async function createBrowserServer(
  project: WorkspaceProject,
  configFile: string | undefined,
  prePlugins: Plugin[] = [],
  postPlugins: Plugin[] = [],
) {
  if (project.ctx.version !== version) {
    project.ctx.logger.warn(
      c.yellow(
        `Loaded ${c.inverse(c.yellow(` vitest@${project.ctx.version} `))} and ${c.inverse(c.yellow(` @vitest/browser@${version} `))}.`
        + '\nRunning mixed versions is not supported and may lead into bugs'
        + '\nUpdate your dependencies and make sure the versions match.',
      ),
    )
  }

  const server = new BrowserServer(project, '/')

  const configPath = typeof configFile === 'string' ? configFile : false

  const logLevel = (process.env.VITEST_BROWSER_DEBUG as 'info') ?? 'info'

  const logger = createViteLogger(logLevel)

  const vite = await createViteServer({
    ...project.options, // spread project config inlined in root workspace config
    base: '/',
    logLevel,
    customLogger: {
      ...logger,
      info(msg, options) {
        logger.info(msg, options)
        if (msg.includes('optimized dependencies changed. reloading')) {
          logger.warn(
            [
              c.yellow(`\n${c.bold('[vitest]')} Vite unexpectedly reloaded a test. This may cause tests to fail, lead to flaky behaviour or duplicated test runs.\n`),
              c.yellow(`For a stable experience, please add mentioned dependencies to your config\'s ${c.bold('\`optimizeDeps.include\`')} field manually.\n\n`),
            ].join(''),
          )
        }
      },
    },
    mode: project.config.mode,
    configFile: configPath,
    // watch is handled by Vitest
    server: {
      hmr: false,
      watch: null,
    },
    plugins: [
      ...prePlugins,
      ...(project.options?.plugins || []),
      BrowserPlugin(server),
      ...postPlugins,
    ],
  })

  await vite.listen()

  setupBrowserRpc(server)

  return server
}
