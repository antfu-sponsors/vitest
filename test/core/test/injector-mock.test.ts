import { Parser } from 'acorn'
import { injectVitestModule } from 'vitest/src/node/esmInjector'
import { expect, test } from 'vitest'

function parse(code: string, options: any) {
  return Parser.parse(code, options)
}

function injectSimpleCode(code: string) {
  return injectVitestModule(code, '/test.js', parse, {
    hijackESM: false,
    cacheDir: '/tmp',
  })?.code.trim()
}

function injectHijackedCode(code: string) {
  return injectVitestModule(code, '/test.js', parse, {
    hijackESM: true,
    cacheDir: '/tmp',
  })?.code.trim()
}

test('hoists mock, unmock, hoisted', () => {
  expect(injectSimpleCode(`
  vi.mock('path', () => {})
  vi.unmock('path')
  vi.hoisted(() => {})
  `)).toMatchInlineSnapshot(`
    "if (typeof globalThis.vi === \\"undefined\\" && typeof globalThis.vitest === \\"undefined\\") { throw new Error(\\"There are some problems in resolving the mocks API.\\\\nYou may encounter this issue when importing the mocks API from another module other than 'vitest'.\\\\nTo fix this issue you can either:\\\\n- import the mocks API directly from 'vitest'\\\\n- enable the 'globals' options\\") }
    vi.mock('path', () => {})
    vi.unmock('path')
    vi.hoisted(() => {})"
  `)
})

test('always hoists import from vitest', () => {
  expect(injectSimpleCode(`
  import { vi } from 'vitest'
  vi.mock('path', () => {})
  vi.unmock('path')
  vi.hoisted(() => {})
  import { test } from 'vitest'
  `)).toMatchInlineSnapshot(`
    "import { vi } from 'vitest'
    import { test } from 'vitest'
    vi.mock('path', () => {})
    vi.unmock('path')
    vi.hoisted(() => {})"
  `)
})

test('always hoists mock, unmock, hoisted when modules are hijacked', () => {
  expect(injectHijackedCode(`
  import { vi } from 'vitest'
  vi.mock('path', () => {})
  vi.unmock('path')
  vi.hoisted(() => {})
  import { test } from 'vitest'
  `)).toMatchInlineSnapshot(`
    "import { __vi_inject__ as __vi_esm_0__ } from 'vitest'
    const { vi } = __vi_esm_0__;
    import { __vi_inject__ as __vi_esm_1__ } from 'vitest'
    const { test } = __vi_esm_1__;
    vi.mock('path', () => {})
    vi.unmock('path')
    vi.hoisted(() => {})"
  `)
})

test('always hoists all imports but they are under mocks', () => {
  expect(injectSimpleCode(`
  import { vi } from 'vitest'
  import { someValue } from './path.js'
  import { someValue2 } from './path2.js'
  vi.mock('path', () => {})
  vi.unmock('path')
  vi.hoisted(() => {})
  import { test } from 'vitest'
  `)).toMatchInlineSnapshot(`
    "import { vi } from 'vitest'
    import { test } from 'vitest'
    vi.mock('path', () => {})
    vi.unmock('path')
    vi.hoisted(() => {})
    const { someValue } = await import('./path.js')
    const { someValue2 } = await import('./path2.js')"
  `)
})

test('always hoists all imports but they are under mocks when modules are hijacked', () => {
  expect(injectHijackedCode(`
  import { vi } from 'vitest'
  import { someValue } from './path.js'
  import { someValue2 } from './path2.js'
  vi.mock('path', () => {})
  vi.unmock('path')
  vi.hoisted(() => {})
  import { test } from 'vitest'
  `)).toMatchInlineSnapshot(`
    "import { __vi_inject__ as __vi_esm_0__ } from 'vitest'
    const { vi } = __vi_esm_0__;
    import { __vi_inject__ as __vi_esm_3__ } from 'vitest'
    const { test } = __vi_esm_3__;
    vi.mock('path', () => {})
    vi.unmock('path')
    vi.hoisted(() => {})
    const { __vi_inject__: __vi_esm_1__ } = await __vi_wrap_module__(import('./path.js'))
    const { __vi_inject__: __vi_esm_2__ } = await __vi_wrap_module__(import('./path2.js'))"
  `)
})
