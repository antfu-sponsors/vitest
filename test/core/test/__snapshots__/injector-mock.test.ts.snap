// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`throws an error when nodes are incompatible > correctly throws an error 1`] = `"Cannot call vi.hoisted() inside vi.mock(): both methods are hoisted to the top of the file and not actually called inside each other."`;

exports[`throws an error when nodes are incompatible > correctly throws an error 2`] = `
"      3|     
      4|     vi.mock('./mocked', () => {
      5|       const variable = vi.hoisted(() => 1)
       |                        ^
      6|       console.log(variable)
      7|     })"
`;

exports[`throws an error when nodes are incompatible > correctly throws an error 3`] = `"Cannot call vi.hoisted() inside vi.mock(): both methods are hoisted to the top of the file and not actually called inside each other."`;

exports[`throws an error when nodes are incompatible > correctly throws an error 4`] = `
"      3| 
      4| vi.mock('./mocked', async () => {
      5|   await vi.hoisted(() => 1)
       |         ^
      6| })
      7|     "
`;

exports[`throws an error when nodes are incompatible > correctly throws an error 5`] = `"Cannot call vi.hoisted() inside vi.mock(): both methods are hoisted to the top of the file and not actually called inside each other."`;

exports[`throws an error when nodes are incompatible > correctly throws an error 6`] = `
"      3| 
      4| vi.mock('./mocked', async () => {
      5|   const variable = await vi.hoisted(() => 1)
       |                          ^
      6| })
      7|     "
`;

exports[`throws an error when nodes are incompatible > correctly throws an error 7`] = `"Cannot call vi.mock() inside vi.hoisted(): both methods are hoisted to the top of the file and not actually called inside each other."`;

exports[`throws an error when nodes are incompatible > correctly throws an error 8`] = `
"      3| 
      4| vi.hoisted(() => {
      5|   vi.mock('./mocked')
       |   ^
      6| })
      7|     "
`;

exports[`throws an error when nodes are incompatible > correctly throws an error 9`] = `"Cannot call vi.mock() inside vi.hoisted(): both methods are hoisted to the top of the file and not actually called inside each other."`;

exports[`throws an error when nodes are incompatible > correctly throws an error 10`] = `
"      3| 
      4| const values = vi.hoisted(() => {
      5|   vi.mock('./mocked')
       |   ^
      6| })
      7|     "
`;

exports[`throws an error when nodes are incompatible > correctly throws an error 11`] = `"Cannot call vi.mock() inside vi.hoisted(): both methods are hoisted to the top of the file and not actually called inside each other."`;

exports[`throws an error when nodes are incompatible > correctly throws an error 12`] = `
"      3| 
      4| await vi.hoisted(async () => {
      5|   vi.mock('./mocked')
       |   ^
      6| })
      7|     "
`;

exports[`throws an error when nodes are incompatible > correctly throws an error 13`] = `"Cannot call vi.mock() inside vi.hoisted(): both methods are hoisted to the top of the file and not actually called inside each other."`;

exports[`throws an error when nodes are incompatible > correctly throws an error 14`] = `
"      3| 
      4| const values = await vi.hoisted(async () => {
      5|   vi.mock('./mocked')
       |   ^
      6| })
      7|     "
`;
